geo/src/algorithm/relate/geomgraph/geometry_graph.rs

```diff
@@ -0,0 +1,387 @@
+/// use composition and delegation to the same effect.
+impl<F> GeometryGraph<'_, F>
+where
+    F: GeoFloat,
+{
+    pub fn edges(&self) -> &[Rc<RefCell<Edge<F>>>] {
+        self.planar_graph.edges()
+    }
+
+    pub fn insert_edge(&mut self, edge: Edge<F>) {
+        self.planar_graph.insert_edge(edge)
+    }
+
+    pub fn is_boundary_node(&self, coord: Coordinate<F>) -> bool {
+        self.planar_graph.is_boundary_node(self.arg_index, coord)
+    }
+
+    pub fn add_node_with_coordinate(&mut self, coord: Coordinate<F>) -> &mut CoordNode<F> {
+        self.planar_graph.add_node_with_coordinate(coord)
+    }
+
+    pub fn nodes_iter(&self) -> impl Iterator<Item = &CoordNode<F>> {
+        self.planar_graph.nodes.iter()
+    }
+}
+
+impl<'a, F> GeometryGraph<'a, F>
+where
+    F: GeoFloat,
+{
+    pub fn new(arg_index: usize, parent_geometry: &'a GeometryCow<F>) -> Self {
+        let mut graph = GeometryGraph {
+            arg_index,
+            parent_geometry,
+            use_boundary_determination_rule: true,
+            planar_graph: PlanarGraph::new(),
+        };
+        graph.add_geometry(parent_geometry);
+        graph
+    }
+
+    pub fn geometry(&self) -> &GeometryCow<F> {
+        self.parent_geometry
+    }
+
+    /// Determine whether a component (node or edge) that appears multiple times in elements
+    /// of a Multi-Geometry is in the boundary or the interior of the Geometry
+    pub fn determine_boundary(boundary_count: usize) -> CoordPos {
+        // For now, we only support the SFS "Mod-2 Rule"
+        // We could make this configurable if we wanted to support alternative boundary rules.
+        if boundary_count % 2 == 1 {
+            CoordPos::OnBoundary
+        } else {
+            CoordPos::Inside
+        }
+    }
+
+    fn create_edge_set_intersector() -> Box<dyn EdgeSetIntersector<F>> {
+        // PERF: faster algorithms exist. This one was chosen for simplicity of implementation and
+        //       debugging
+        Box::new(SimpleEdgeSetIntersector::new())
+    }
+
+    fn boundary_nodes(&self) -> impl Iterator<Item = &CoordNode<F>> {
+        self.planar_graph.boundary_nodes(self.arg_index)
+    }
+
+    pub fn add_geometry(&mut self, geometry: &GeometryCow<F>) {
+        if geometry.is_empty() {
+            return;
+        }
+        match geometry {
+            GeometryCow::Line(line) => self.add_line(line),
+            GeometryCow::Rect(rect) => {
+                // PERF: avoid this conversion/clone?
+                self.add_polygon(&rect.to_polygon());
+            }
+            GeometryCow::Point(point) => {
+                self.add_point(point);
+            }
+            GeometryCow::Polygon(polygon) => self.add_polygon(polygon),
+            GeometryCow::Triangle(triangle) => {
+                // PERF: avoid this conversion/clone?
+                self.add_polygon(&Polygon::from(triangle.clone().into_owned()));
```

Comment:
is the `into_owned()` here redundant with `clone`?

Reply:
Good catch! 

I've changed this to `triangle.to_polygon()` which I think is still clearer.